### PR TITLE
Muonrecofix

### PR DESCRIPTION
**Pull request by Sojourner92**
Fix for issue #603. Most notable changes:

1. Added primary mirror hole radii to be taken in to consideration in the muon reconstruction for every telescope to muon_reco_functions.py (line 45), based on the latest MC parameter description files I could find.
2. Fixed an issue in muon_integrator.py arising due to non-astropy quantity objects being assigned astropy quantities.
3. Fixed an bug in the muon_reconstruction.py example script where detected muon rings were not being correctly logged.

As far as I can tell, this works properly now, but an expert in muon analysis (which I'm not) might want to double check the final results thrown out by muon_reconstruction.py to make sure they're sensible.

### CODE REVIEW

File: ctapipe/image/muon/muon_integrator.py [hunk 1]
```diff
@@ -68,7 +68,8 @@ def __init__(self, mirror_radius, hole_radius, pixel_width=0.2,
         self.prediction = 0
         self.minlambda = 300.e-9
         self.maxlambda = 600.e-9
-        self.photemit = alpha * (self.minlambda**-1 - self.maxlambda**-1)  # 12165.45
+        self.photemit = 1.0 / 137.0 * (self.minlambda**-1 -
```
Comment (MaxNoe) [inline ctapipe/image/muon/muon_integrator.py]: Why replace a well-known constant with two magic numbers?

Comment (kosack) [reply ctapipe/image/muon/muon_integrator.py]: at least use `scipy.constants.alpha` (or `scipy.constants.inverse_alpha`)

Comment (kosack) [reply ctapipe/image/muon/muon_integrator.py]: or `astropy.constants.alpha` which is also there, and has references to where the value comes from automatically

File: ctapipe/image/muon/muon_integrator.py [hunk 2]
```diff
@@ -68,7 +68,8 @@ def __init__(self, mirror_radius, hole_radius, pixel_width=0.2,
         self.prediction = 0
         self.minlambda = 300.e-9
```
Comment (kosack) [inline ctapipe/image/muon/muon_integrator.py]: if these are wavelengths, please require them to be astropy `Quantities`, so the unit is attached (e.g. define them as `300e-9 * u.m` or even `30 * u.angstrom`.

**Comment by rlopezcoto** [comment]
Hi, 
I checked your fix but I got the error:

```sh
Traceback (most recent call last):
  File "muon_reconstruction.py", line 134, in <module>
    main()
  File "muon_reconstruction.py", line 91, in main
    muon_evt = analyze_muon_event(event)
  File "/home/rlopez/data_rlopez/cta/ctasoft/test/ctapipe/ctapipe/image/muon/muon_reco_functions.py", line 93, in analyze_muon_event
    dict_index = muon_cuts['Name'].index(str(teldes))
ValueError: 'SST-ASTRI:CHEC' is not in list
```

if I add 'SST-ASTRI:CHEC' to the dictionary entries in **muon_reco_functions.py** and use the same TailCuts, TotalPix... as in the 'SST-GCT:CHEC', that error is gone, but now I have an error related to the baseframe coordinates:

```sh
Traceback (most recent call last):
  File "muon_reconstruction.py", line 134, in <module>
    main()
  File "muon_reconstruction.py", line 91, in main
    muon_evt = analyze_muon_event(event)
  File "/home/rlopez/data_rlopez/cta/ctasoft/test/ctapipe/ctapipe/image/muon/muon_reco_functions.py", line 110, in analyze_muon_event
    rotation=geom.pix_rotation)
  File "/home/rlopez/data_rlopez/software/anaconda3/envs/cta-dev/lib/python3.6/site-packages/astropy/coordinates/baseframe.py", line 405, in __init__
    list(kwargs)))
TypeError: Coordinate frame got unexpected keywords: ['z']
```
do you get anything similar?

**Comment by Sojourner92** [comment]
Hi,
No I haven't encountered anything similar (I think the script predates the astri/chec simulation studies), but if you post a simtelarray file I'll try to reproduce your error and see if I can fix it.

**Comment by rlopezcoto** [comment]
Hi,
I tried with different files, but I get it just running the test file in the example:
ctapipe-extra/ctapipe_resources/gamma_test.simtel.gz

```sh
$python muon_reconstruction.py 
INFO: [SCRIPT] muon_reconstruction [__main__]
INFO:astropy:[SCRIPT] muon_reconstruction
Doing one-time thread-specific warning data initialization.
I/O block extended by 744748 to 1744748 bytes
Event Number 0
Traceback (most recent call last):
  File "muon_reconstruction.py", line 134, in <module>
    main()
  File "muon_reconstruction.py", line 91, in main
    muon_evt = analyze_muon_event(event)
  File "/home/rlopez/data_rlopez/cta/ctasoft/test/ctapipe/ctapipe/image/muon/muon_reco_functions.py", line 93, in analyze_muon_event
    dict_index = muon_cuts['Name'].index(str(teldes))
ValueError: 'SST-ASTRI:CHEC' is not in list

gzip: stdout: Broken pipe
```

as I said, to solve this is easy just by adding the 'SST-ASTRI:CHEC' dictionary entry, but then you have to solve mor problems and I wanted to know if you had encounter anything similar before digging deeper. Is it possible that you have an older gamma_test.simtel.gz file? I just checked it out from the repo.

**Comment by Sojourner92** [comment]
Yeah, I wasn't running it with the example script, I was running the MuonDisplayerTool function with the data provided in #603 for NectarCAM and ipython. However, I don't get the same crash as you doing the same for the example file, albeit as I'm using the ASTRI parameters for the mirror radius etc as these will be more appropriate for an ASTRI:CHEC run. The parameters I'm using when I add the ASTRI: CHEC camera to the dictionary are:

Tailcuts=(5,7)
impact=(0.1,0.95)
Totalpix=2048
Minpix=164
AngPixelWidth=0.2
hole_rad=0.171 * u.m
cam_rad=2.86
sec_rad=1.8 * u.m
sct=True

I don't get a crash, but the tool isn't finding any muons, I think probably because the angular pixel width isn't correct. Does anyone know what value this should take for CHEC on ASTRI? There isn't a MC parameter description for that configuration.

**Comment by rlopezcoto** [comment]
I run it with the parameters you wrote up there, but I still get the Coordinate frame error.
If the tool is not finding any muons, it might be because the example file is a gamma file.

**Comment by rlopezcoto** [comment]
I found a problem that seems related to the rotation of the fitted circle, at least when plotted with the --display option (see images). 
screen shot 2018-01-22 at 19 43 00
screen shot 2018-01-22 at 19 49 25
screen shot 2018-01-22 at 19 49 47

It does not look like it is affecting the width measurement and maybe either the other muon measured parameters, so it is probably a rotation problem. If you want I can share with you a muon file that @AMWMitchell sent me.
Sorry, not sure if this bug is related to this PR or it was already there before, I'm reporting it here just in case, otherwise I can open an issue and investigate it.

**Comment by AMWMitchell** [comment]
Not again - had this issue before. It looks like the definition of rotation angle has changed again. 
See plot_muon_event in muon_diagnostic_plots and try commentating out the if statement on LSTCam / NectarCam concerning the rotation angle. 
(might need to check how geom.pix_rotation is defined)

**Comment by watsonjj** [comment]
Once this has been fixed, is there a test that can be created to check that this isn't broken again with a future update?

**Comment by AMWMitchell** [comment]
Good point - although no easy tests spring to mind other than visually. 
Might have to do something like check that the pixels that pass the ring distance cut (pix distance to ring radius) in the nominal co-ordinate system also pass the same cut in the Camera co-ordinate system (using transformed ring centroid etc.)

**Comment by rlopezcoto** [comment]
Commenting out lines 129 and 131 of muon_diagnostic_plots.py solved the issue. @Sojourner92, you may want to include it in your PR and we can maybe think later on a test to check this. 
screen shot 2018-01-23 at 08 19 02

**Comment by kosack** [comment]
Is this ready to be merged, or is there more work being added?

**Comment by Sojourner92** [comment]
Hi Karl,
No not yet, I've started having problems since I pulled the latest upstream master version. The camera dl1 image.charge_extractors modules are trying to import a function that no longer appears to exist on line 14 , 
from ctapipe.utils.neighbour_sum import get_sum_array.

Does anyone know anything about changes to this? It seems as though a python function has been replaced by a c++ wrapper, but I'm not sure how to fix the reference in the charge extractor file.

**Comment by watsonjj** [comment]
That functionality has not be changed in the recent PRs. I suggest trying `python setup.py develop` to force a recompile of the c code

**Comment by Sojourner92** [comment]
Thanks Jason, that appears to have fixed everything. Its now ready for merging then Karl.

**Pull request merged**